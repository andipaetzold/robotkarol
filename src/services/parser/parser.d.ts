import { ParseErrorData } from "./ParseError";
import { AST } from "./types";

declare module "./parser.js" {
  export class Parser {
    constructor();

    yy: {
      parseError?: (message: string, data: ParseErrorData) => void;
    };

    parse(code: string): AST;
  }

  export function parse(code: string): AST;
}
