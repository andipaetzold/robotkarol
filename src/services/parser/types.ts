export type AST = {
  type: "ast";
  program: ASTProgram;
  functions: ASTFunction[];
};

export interface ASTProgram {
  type: "program";
  line: number;
  body: ASTStatement[];
}

export interface ASTFunction {
  type: "function";
  line: number;
  identifier: string;
  body: ASTStatements;
}

export type ASTStatements = ASTStatement[];
export type ASTStatement =
  | ASTIfStatement
  | ASTCall
  | ASTSystemCall
  | ASTFunctionCall
  | ASTRepeatStatement
  | ASTWhileStatement;

export interface ASTCall {
  type: "call";
  line: number;
  action: Action;
}

export interface ASTSystemCall {
  type: "systemCall";
  line: number;
  action: SystemAction;
}

export interface ASTFunctionCall {
  type: "functionCall";
  line: number;
  name: string;
}

export interface ASTRepeatStatement {
  type: "repeat";
  line: number;
  times: number;
  body: ASTStatements;
}

export interface ASTWhileStatement {
  type: "while";
  line: number;
  condition: ASTCondition;
  body: ASTStatements;
}

export interface ASTIfStatement {
  type: "if";
  line: number;
  condition: ASTCondition;
  body: ASTStatements;
  elseBody: ASTStatements;
}

export type ASTCondition = ASTExpression | ASTNotExpression;

export interface ASTNotExpression {
  type: "not";
  line: number;
  condition: ASTCondition;
}

export interface ASTExpression {
  type: "expression";
  line: number;
  test: Test;
}

export type Test =
  | "IS_WALL"
  | "NOT_IS_WALL"
  | "IS_BRICK"
  | "NOT_IS_BRICK"
  | "IS_MARKER"
  | "NOT_IS_MARKER"
  | "IS_NORTH"
  | "IS_EAST"
  | "IS_SOUTH"
  | "IS_WEST";

export type Action =
  | "STEP"
  | "TURN_LEFT"
  | "TURN_RIGHT"
  | "BRICK_PUT"
  | "BRICK_TAKE"
  | "MARKER_SET"
  | "MARKER_REMOVE";

export type SystemAction = "fast" | "slow";
