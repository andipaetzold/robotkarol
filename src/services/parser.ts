import { Parser } from "jison";
import { AST } from "../types";

const grammar = {
  lex: {
    options: {
      "case-insensitive": true,
    },
    rules: [
      ["\\s+", "/* skip whitespace */"],
      ["\\{[^\\}]+\\}", "/* skip comments */"],
      ["programm\\b", "return 'PROGRAM_BEGIN'"],
      ["\\*programm\\b", "return 'PROGRAM_END'"],
      ["endeprogramm\\b", "return 'PROGRAM_END'"],
      ["anweisung\\b", "return 'FUNCTION_BEGIN'"],
      ["\\*anweisung\\b", "return 'FUNCTION_END'"],
      ["endeanweisung\\b", "return 'FUNCTION_END'"],
      ["schritt\\b", "return 'STEP'"],
      ["linksdrehen\\b", "return 'TURN_LEFT'"],
      ["rechtsdrehen\\b", "return 'TURN_RIGHT'"],
      ["hinlegen\\b", "return 'BRICK_PUT'"],
      ["aufheben\\b", "return 'BRICK_TAKE'"],
      ["markesetzen\\b", "return 'MARKER_SET'"],
      ["markelöschen\\b", "return 'MARKER_REMOVE'"],
      ["wiederhole\\b", "return 'REPEAT_BEGIN'"],
      ["\\*wiederhole\\b", "return 'REPEAT_END'"],
      ["solange\\b", "return 'WHILE'"],
      ["mal\\b", "return 'TIMES'"],
      ["wenn\\b", "return 'IF'"],
      ["dann\\b", "return 'THEN'"],
      ["endewenn\\b", "return 'IF_END'"],
      ["\\*wenn\\b", "return 'IF_END'"],
      ["sonst\\b", "return 'ELSE'"],
      ["istwand\\b", "return 'IS_WALL'"],
      ["nichtistwand\\b", "return 'NOT_IS_WALL'"],
      ["istziegel\\b", "return 'IS_BRICK'"],
      ["nichtistziegel\\b", "return 'NOT_IS_BRICK'"],
      ["istnorden\\b", "return 'IS_NORTH'"],
      ["istosten\\b", "return 'IS_EAST'"],
      ["istsüden\\b", "return 'IS_SOUTH'"],
      ["istwesten\\b", "return 'IS_WEST'"],
      ["nicht\\b", "return 'NOT'"],
      ["istmarke\\b", "return 'IS_MARKER'"],
      ["nichtistmarke\\b", "return 'NOT_IS_MARKER'"],
      ["[0-9]+(?:\\.[0-9]+)?\\b", "return 'NUMBER'"],
      ["[a-z]+\\b", "return 'IDENTIFIER'"],
    ],
  },

  bnf: {
    start: [["segments", "return $1"]],
    segments: [
      ["segment", "$$ = $1"],
      ["segment segments", "$$ = $1 + $2"],
    ],
    segment: [
      [
        "PROGRAM_BEGIN statements PROGRAM_END",
        "$$ = { type: 'program', body: $2 }",
      ],
      ["PROGRAM_BEGIN PROGRAM_END", "$$ = { type: 'program', body: [] }"],
      [
        "FUNCTION_BEGIN IDENTIFIER statements FUNCTION_END",
        "$$ = { type: 'function', identifier: $2, body: $3 }",
      ],
      [
        "FUNCTION_BEGIN IDENTIFIER FUNCTION_END",
        "$$ = { type: 'function', identifier: $2, body: [] }",
      ],
      ,
    ],
    statements: [
      ["statement", "$$ = [$1]"],
      ["statement statements", "$$ = [$1, ...$2]"],
    ],
    statement: [
      ["STEP", "$$ = { type: 'call', action: 'STEP' }"],
      ["TURN_LEFT", "$$ = { type: 'call', action: 'TURN_LEFT' }"],
      ["TURN_RIGHT", "$$ = { type: 'call', action: 'TURN_RIGHT' }"],
      ["BRICK_PUT", "$$ = { type: 'call', action: 'BRICK_PUT' }"],
      ["BRICK_TAKE", "$$ = { type: 'call', action: 'BRICK_TAKE' }"],
      ["MARKER_SET", "$$ = { type: 'call', action: 'MARKER_SET' }"],
      ["MARKER_REMOVE", "$$ = { type: 'call', action: 'MARKER_REMOVE' }"],
      [
        "REPEAT_BEGIN NUMBER TIMES statements REPEAT_END",
        "$$ = { type: 'repeat', times: $2, body: $4 }",
      ],
      [
        "REPEAT_BEGIN WHILE condition statements REPEAT_END",
        "$$ = { type: 'while', condition: $3, body: $4 }",
      ],
      [
        "IF condition THEN statements IF_END",
        "$$ = { type: 'if', condition: $2, body: $4, elseBody: [] }",
      ],
      [
        "IF condition THEN statements ELSE statements IF_END",
        "$$ = { type: 'if', condition: $2, body: $4, elseBody: $6 }",
      ],
    ],
    condition: [
      ["NOT condition", "$$ = { type: 'not', expression: $2 }"],
      ["IS_WALL", "$$ = { type: 'expression', test: 'IS_WALL' }"],
      ["NOT_IS_WALL", "$$ = { type: 'expression', test: 'NOT_IS_WALL' }"],
      ["IS_BRICK", "$$ = { type: 'expression', test: 'IS_BRICK' }"],
      ["NOT_IS_BRICK", "$$ = { type: 'expression', test: 'NOT_IS_BRICK' }"],
      ["IS_MARKER", "$$ = { type: 'expression', test: 'IS_MARKER' }"],
      ["NOT_IS_MARKER", "$$ = { type: 'expression', test: 'NOT_IS_MARKER' }"],
      ["IS_NORTH", "$$ = { type: 'expression', test: 'IS_NORTH' }"],
      ["IS_WEST", "$$ = { type: 'expression', test: 'IS_WEST' }"],
      ["IS_SOUTH", "$$ = { type: 'expression', test: 'IS_SOUTH' }"],
      ["IS_EAST", "$$ = { type: 'expression', test: 'IS_EAST' }"],
    ],
  },
};

const parser = new Parser(grammar);

export function parse(code: string): AST {
  return parser.parse(code);
}
