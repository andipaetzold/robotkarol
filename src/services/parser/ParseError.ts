export type ParseErrorData = {
  expected?: string[];
  line?: number;
  loc?: {
    first_column: number;
    first_row: number;
    last_column: number;
    last_line: number;
  };
  text?: string;
  token?: number | string | null;
};

export class ParseError extends Error {
  constructor(message: string, public readonly data: ParseErrorData = {}) {
    super(message);
  }
}
