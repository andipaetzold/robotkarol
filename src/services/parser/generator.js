const { Parser } = require("jison");
const { writeFileSync } = require("fs");

const grammar = {
  lex: {
    options: {
      "case-insensitive": true,
    },
    rules: [
      ["\\s+", "/* skip whitespace */"],

      ["\\{[^\\}]*\\}", "/* skip multiline comments */"],
      ["\\/\\/[^\\n]*", "/* skip single line comments */"],

      ["programm\\b", "return 'PROGRAM_BEGIN'"],
      ["(\\*programm|endeprogramm)\\b", "return 'PROGRAM_END'"],
      ["anweisung\\b", "return 'FUNCTION_BEGIN'"],
      ["(\\*anweisung|endeanweisung)\\b", "return 'FUNCTION_END'"],

      ["schritt\\b", "return 'STEP'"],
      ["linksdrehen\\b", "return 'TURN_LEFT'"],
      ["rechtsdrehen\\b", "return 'TURN_RIGHT'"],
      ["hinlegen\\b", "return 'BRICK_PUT'"],
      ["aufheben\\b", "return 'BRICK_TAKE'"],
      ["markesetzen\\b", "return 'MARKER_SET'"],
      ["markelöschen\\b", "return 'MARKER_REMOVE'"],

      ["wiederhole\\b", "return 'REPEAT_BEGIN'"],
      ["(\\*wiederhole|endewiederhole)\\b", "return 'REPEAT_END'"],
      ["immer\\b", "return 'ALWAYS'"],
      ["solange\\b", "return 'WHILE'"],
      ["(\\*solange|endesolange)\\b", "return 'WHILE_END'"],
      ["mal\\b", "return 'TIMES'"],

      ["wenn\\b", "return 'IF'"],
      ["dann\\b", "return 'THEN'"],
      ["(\\*wenn|endewenn)\\b", "return 'IF_END'"],
      ["sonst\\b", "return 'ELSE'"],

      ["istwand\\b", "return 'IS_WALL'"],
      ["nichtistwand\\b", "return 'NOT_IS_WALL'"],
      ["istziegel\\b", "return 'IS_BRICK'"],
      ["nichtistziegel\\b", "return 'NOT_IS_BRICK'"],
      ["istnorden\\b", "return 'IS_NORTH'"],
      ["istosten\\b", "return 'IS_EAST'"],
      ["istsüden\\b", "return 'IS_SOUTH'"],
      ["istwesten\\b", "return 'IS_WEST'"],
      ["nicht\\b", "return 'NOT'"],
      ["istmarke\\b", "return 'IS_MARKER'"],
      ["nichtistmarke\\b", "return 'NOT_IS_MARKER'"],

      ["schnell\\b", "return 'FAST'"],
      ["langsam\\b", "return 'SLOW'"],
      ["beenden\\b", "return 'EXIT'"],
      ["ton\\b", "return 'SOUND'"],

      ["[0-9]+(?:\\.[0-9]+)?\\b", "return 'NUMBER'"],
      ["[a-z0-9_\\-äöüß]+\\b", "return 'IDENTIFIER'"],
    ],
  },

  bnf: {
    start: [
      [
        "segments",
        `return {
      type: 'ast',
      functions: $1.filter(s => s.type === 'function'),
      program: $1.find(s => s.type === 'program') ??
        {
          type: 'program',
          line: $1.filter(s => s.type !== 'function').filter(s => s.type !== 'program')[0]?.line ?? 0,
          body: $1.filter(s => s.type !== 'function').filter(s => s.type !== 'program')
        }
    }`,
      ],
    ],
    segments: [
      ["segment", "$$ = [$1]"],
      ["segment segments", "$$ = [$1, ...$2]"],
    ],
    segment: [
      ["program", "$$ = $1"],
      ["function", "$$ = $1"],
      ["statement", "$$ = $1"],
    ],
    program: [
      [
        "PROGRAM_BEGIN statements PROGRAM_END",
        "$$ = { type: 'program', line: yylineno, body: $2 }",
      ],
      [
        "PROGRAM_BEGIN PROGRAM_END",
        "$$ = { type: 'program', line: yylineno, body: [] }",
      ],
    ],
    function: [
      [
        "FUNCTION_BEGIN IDENTIFIER statements FUNCTION_END",
        "$$ = { type: 'function', line: yylineno, identifier: $2, body: $3 }",
      ],
      [
        "FUNCTION_BEGIN IDENTIFIER FUNCTION_END",
        "$$ = { type: 'function', line: yylineno, identifier: $2, body: [] }",
      ],
    ],
    statements: [
      ["statement", "$$ = [$1]"],
      ["statement statements", "$$ = [$1, ...$2]"],
    ],
    statement: [
      ["STEP", "$$ = { type: 'call', line: yylineno, action: 'STEP' }"],
      [
        "TURN_LEFT",
        "$$ = { type: 'call', line: yylineno, action: 'TURN_LEFT' }",
      ],
      [
        "TURN_RIGHT",
        "$$ = { type: 'call', line: yylineno, action: 'TURN_RIGHT' }",
      ],
      [
        "BRICK_PUT",
        "$$ = { type: 'call', line: yylineno, action: 'BRICK_PUT' }",
      ],
      [
        "BRICK_TAKE",
        "$$ = { type: 'call', line: yylineno, action: 'BRICK_TAKE' }",
      ],
      [
        "MARKER_SET",
        "$$ = { type: 'call', line: yylineno, action: 'MARKER_SET' }",
      ],
      [
        "MARKER_REMOVE",
        "$$ = { type: 'call', line: yylineno, action: 'MARKER_REMOVE' }",
      ],

      ["SLOW", "$$ = { type: 'systemCall', line: yylineno, action: 'slow' }"],
      ["FAST", "$$ = { type: 'systemCall', line: yylineno, action: 'fast' }"],
      ["EXIT", "$$ = { type: 'systemCall', line: yylineno, action: 'exit' }"],
      ["SOUND", "$$ = { type: 'systemCall', line: yylineno, action: 'sound' }"],

      [
        "REPEAT_BEGIN NUMBER TIMES statements REPEAT_END",
        "$$ = { type: 'repeat', line: yylineno, times: $2, body: $4 }",
      ],
      [
        "REPEAT_BEGIN ALWAYS statements REPEAT_END",
        "$$ = { type: 'repeat', line: yylineno, times: Infinity, body: $3 }",
      ],
      [
        "REPEAT_BEGIN WHILE condition statements REPEAT_END",
        "$$ = { type: 'while', line: yylineno, condition: $3, body: $4 }",
      ],
      [
        "WHILE condition statements WHILE_END",
        "$$ = { type: 'while', line: yylineno, condition: $3, body: $4 }",
      ],

      [
        "IF condition THEN statements IF_END",
        "$$ = { type: 'if', line: yylineno, condition: $2, body: $4, elseBody: [] }",
      ],
      [
        "IF condition THEN statements ELSE statements IF_END",
        "$$ = { type: 'if', line: yylineno, condition: $2, body: $4, elseBody: $6 }",
      ],

      ["IDENTIFIER", "$$ = { type: 'functionCall', name: $1, line: yylineno }"],
    ],
    condition: [
      ["NOT condition", "$$ = { type: 'not', line: yylineno, condition: $2 }"],
      [
        "IS_WALL",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_WALL' }",
      ],
      [
        "NOT_IS_WALL",
        "$$ = { type: 'expression', line: yylineno, test: 'NOT_IS_WALL' }",
      ],
      [
        "IS_BRICK",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_BRICK' }",
      ],
      [
        "NOT_IS_BRICK",
        "$$ = { type: 'expression', line: yylineno, test: 'NOT_IS_BRICK' }",
      ],
      [
        "IS_MARKER",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_MARKER' }",
      ],
      [
        "NOT_IS_MARKER",
        "$$ = { type: 'expression', line: yylineno, test: 'NOT_IS_MARKER' }",
      ],
      [
        "IS_NORTH",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_NORTH' }",
      ],
      [
        "IS_WEST",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_WEST' }",
      ],
      [
        "IS_SOUTH",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_SOUTH' }",
      ],
      [
        "IS_EAST",
        "$$ = { type: 'expression', line: yylineno, test: 'IS_EAST' }",
      ],
    ],
  },
};

const parser = new Parser(grammar);
const code = parser.generate({ moduleType: "js" });
writeFileSync(
  `${__dirname}/parser.js`,
  `${code}
export const Parser = parser.Parser;
export const parse = function () { return parser.parse.apply(parser, arguments); };`
);
