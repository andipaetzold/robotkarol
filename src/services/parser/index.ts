import { groupBy } from "lodash";
import { ParseError, ParseErrorData } from "./ParseError";
import { Parser } from "./parser";
import { AST, ASTFunctionCall, ASTStatement } from "./types";

const parser = new Parser();
parser.yy.parseError = (str: string, data: ParseErrorData) => {
  throw new ParseError(str, data);
};

export function parse(code: string): AST {
  const ast = parser.parse(code);
  validate(ast);
  return ast;
}

export function validate(ast: AST) {
  const functionBlocksByName = groupBy(ast.functions, (f) => f.identifier);
  for (const [functionName, blocks] of Object.entries(functionBlocksByName)) {
    if (blocks.length > 1) {
      throw new ParseError(
        `Duplicate function blocks for function '${functionName}'.`,
        { line: blocks[1].line }
      );
    }
  }

  const allFunctionCalls = [...ast.functions, ast.program].flatMap((x) =>
    x.body.flatMap((s) => getFunctionCalls(s))
  );
  const functionNames = ast.functions.map((f) => f.identifier);
  for (const functionCall of allFunctionCalls) {
    if (functionNames.includes(functionCall.name)) {
      continue;
    }

    throw new ParseError(`Function ${functionCall.name} does not exist.`, {
      line: functionCall.line,
    });
  }
}

function getFunctionCalls(statement: ASTStatement): ASTFunctionCall[] {
  switch (statement.type) {
    case "functionCall":
      return [statement];
    case "systemCall":
      return [];
    case "if":
      return [
        ...statement.body.flatMap((s) => getFunctionCalls(s)),
        ...statement.elseBody.flatMap((s) => getFunctionCalls(s)),
      ];
    case "repeat":
    case "while":
      return statement.body.flatMap((s) => getFunctionCalls(s));
    case "call":
      return [];
  }
}
