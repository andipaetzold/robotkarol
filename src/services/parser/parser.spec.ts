import { parse } from "./parser";

describe("program", () => {
  it("should accept empty", () => {
    const code = `
        Programm
        *Programm`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("should accept programs without wrapping `program`", () => {
    const code = `
        Schritt
        Schritt`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("should accept basic body", () => {
    const code = `
        Programm
          Schritt
          Schritt
        *Programm`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });
});

describe("function", () => {
  it("should accept empty", () => {
    const code = `
        Anweisung Test
        *Anweisung`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("should accept basic body", () => {
    const code = `
        Anweisung Test
          Schritt
        *Anweisung`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });
});

it("accept all call commands", () => {
  const code = `
      Schritt
      LinksDrehen
      RechtsDrehen
      Hinlegen
      Aufheben
      MarkeSetzen
      MarkeLÃ¶schen`;
  const ast = parse(code);
  expect(ast).toMatchSnapshot();
});

describe("comments", () => {
  it("ignores multiline", () => {
    const code = `
        { Comment }
        {}
        Programm
        { Comment }
        *Programm
        { Comment }`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("ignores single line", () => {
    const code = `
        // Comment
        //
        Programm
        // Comment
        *Programm
        // Comment`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });
});

describe("if", () => {
  it("all conditions", () => {
    const code = `
      Wenn IstWand Dann
        Schritt
      *Wenn  
      Wenn IstZiegel Dann
        Schritt
      *Wenn
      Wenn IstMarke Dann
        Schritt
      *Wenn
      Wenn IstNorden Dann
        Schritt
      *Wenn`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("negative condition", () => {
    const code = `
        Wenn nicht IstWand Dann
          Schritt
        *Wenn`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("else", () => {
    const code = `
        Wenn IstWand Dann
          LinksDrehen
        Sonst
          RechtsDrehen
        *Wenn`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("nested", () => {
    const code = `
        Wenn IstWand Dann
          LinksDrehen
          Wenn IstZiegel Dann
            Aufheben
          *Wenn
        Sonst
          RechtsDrehen
        *Wenn`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });
});

describe("repeat", () => {
  it("times", () => {
    const code = `
        Wiederhole 5 mal
          Schritt
        *Wiederhole
        
        wiederhole 5 mal
          Schritt
        endewiederhole`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("always", () => {
    const code = `
      wiederhole immer
        Schritt
      *wiederhole`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("while", () => {
    const code = `
        Wiederhole solange IstZiegel
          Aufheben
        *Wiederhole
        
        solange IstZiegel
          Aufheben
        *solange
        
        Wiederhole solange IstZiegel
          Aufheben
        EndeWiederhole`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });
});

describe("system calls", () => {
  it("slow", () => {
    const code = `
        langsam
        Schritt
        Schritt`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });

  it("fast", () => {
    const code = `
        schnell
        Schritt
        Schritt`;
    const ast = parse(code);
    expect(ast).toMatchSnapshot();
  });
});
