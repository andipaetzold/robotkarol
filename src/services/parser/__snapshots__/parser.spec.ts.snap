// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accept all call commands 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "action": "STEP",
        "line": 1,
        "type": "call",
      },
      Object {
        "action": "TURN_LEFT",
        "line": 2,
        "type": "call",
      },
      Object {
        "action": "TURN_RIGHT",
        "line": 3,
        "type": "call",
      },
      Object {
        "action": "BRICK_PUT",
        "line": 4,
        "type": "call",
      },
      Object {
        "action": "BRICK_TAKE",
        "line": 5,
        "type": "call",
      },
      Object {
        "action": "MARKER_SET",
        "line": 6,
        "type": "call",
      },
      Object {
        "action": "MARKER_REMOVE",
        "line": 7,
        "type": "call",
      },
    ],
    "line": 1,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`comments ignores multiline 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [],
    "line": 5,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`comments ignores single line 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [],
    "line": 5,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`function should accept basic body 1`] = `
Object {
  "functions": Array [
    Object {
      "body": Array [
        Object {
          "action": "STEP",
          "line": 2,
          "type": "call",
        },
      ],
      "identifier": "Test",
      "line": 3,
      "type": "function",
    },
  ],
  "program": Object {
    "body": Array [],
    "line": 0,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`function should accept empty 1`] = `
Object {
  "functions": Array [
    Object {
      "body": Array [],
      "identifier": "Test",
      "line": 2,
      "type": "function",
    },
  ],
  "program": Object {
    "body": Array [],
    "line": 0,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`if all conditions 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 2,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 1,
          "test": "IS_WALL",
          "type": "expression",
        },
        "elseBody": Array [],
        "line": 3,
        "type": "if",
      },
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 5,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 4,
          "test": "IS_BRICK",
          "type": "expression",
        },
        "elseBody": Array [],
        "line": 6,
        "type": "if",
      },
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 8,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 7,
          "test": "IS_MARKER",
          "type": "expression",
        },
        "elseBody": Array [],
        "line": 9,
        "type": "if",
      },
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 11,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 10,
          "test": "IS_NORTH",
          "type": "expression",
        },
        "elseBody": Array [],
        "line": 12,
        "type": "if",
      },
    ],
    "line": 3,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`if else 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "TURN_LEFT",
            "line": 2,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 1,
          "test": "IS_WALL",
          "type": "expression",
        },
        "elseBody": Array [
          Object {
            "action": "TURN_RIGHT",
            "line": 4,
            "type": "call",
          },
        ],
        "line": 5,
        "type": "if",
      },
    ],
    "line": 5,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`if negative condition 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 2,
            "type": "call",
          },
        ],
        "condition": Object {
          "condition": Object {
            "line": 1,
            "test": "IS_WALL",
            "type": "expression",
          },
          "line": 1,
          "type": "not",
        },
        "elseBody": Array [],
        "line": 3,
        "type": "if",
      },
    ],
    "line": 3,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`if nested 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "TURN_LEFT",
            "line": 2,
            "type": "call",
          },
          Object {
            "body": Array [
              Object {
                "action": "BRICK_TAKE",
                "line": 4,
                "type": "call",
              },
            ],
            "condition": Object {
              "line": 3,
              "test": "IS_BRICK",
              "type": "expression",
            },
            "elseBody": Array [],
            "line": 5,
            "type": "if",
          },
        ],
        "condition": Object {
          "line": 1,
          "test": "IS_WALL",
          "type": "expression",
        },
        "elseBody": Array [
          Object {
            "action": "TURN_RIGHT",
            "line": 7,
            "type": "call",
          },
        ],
        "line": 8,
        "type": "if",
      },
    ],
    "line": 8,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`program should accept basic body 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "action": "STEP",
        "line": 2,
        "type": "call",
      },
      Object {
        "action": "STEP",
        "line": 3,
        "type": "call",
      },
    ],
    "line": 4,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`program should accept empty 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [],
    "line": 2,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`program should accept programs without wrapping \`program\` 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "action": "STEP",
        "line": 1,
        "type": "call",
      },
      Object {
        "action": "STEP",
        "line": 2,
        "type": "call",
      },
    ],
    "line": 1,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`repeat always 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 2,
            "type": "call",
          },
        ],
        "line": 3,
        "times": Infinity,
        "type": "repeat",
      },
    ],
    "line": 3,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`repeat times 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 2,
            "type": "call",
          },
        ],
        "line": 3,
        "times": "5",
        "type": "repeat",
      },
      Object {
        "body": Array [
          Object {
            "action": "STEP",
            "line": 6,
            "type": "call",
          },
        ],
        "line": 7,
        "times": "5",
        "type": "repeat",
      },
    ],
    "line": 3,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`repeat while 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "body": Array [
          Object {
            "action": "BRICK_TAKE",
            "line": 2,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 1,
          "test": "IS_BRICK",
          "type": "expression",
        },
        "line": 3,
        "type": "while",
      },
      Object {
        "body": "*solange",
        "condition": Array [
          Object {
            "action": "BRICK_TAKE",
            "line": 6,
            "type": "call",
          },
        ],
        "line": 7,
        "type": "while",
      },
      Object {
        "body": Array [
          Object {
            "action": "BRICK_TAKE",
            "line": 10,
            "type": "call",
          },
        ],
        "condition": Object {
          "line": 9,
          "test": "IS_BRICK",
          "type": "expression",
        },
        "line": 11,
        "type": "while",
      },
    ],
    "line": 3,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`system calls fast 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "action": "fast",
        "line": 1,
        "type": "systemCall",
      },
      Object {
        "action": "STEP",
        "line": 2,
        "type": "call",
      },
      Object {
        "action": "STEP",
        "line": 3,
        "type": "call",
      },
    ],
    "line": 1,
    "type": "program",
  },
  "type": "ast",
}
`;

exports[`system calls slow 1`] = `
Object {
  "functions": Array [],
  "program": Object {
    "body": Array [
      Object {
        "action": "slow",
        "line": 1,
        "type": "systemCall",
      },
      Object {
        "action": "STEP",
        "line": 2,
        "type": "call",
      },
      Object {
        "action": "STEP",
        "line": 3,
        "type": "call",
      },
    ],
    "line": 1,
    "type": "program",
  },
  "type": "ast",
}
`;
